File: examples/shared/src/env.ts
import { config } from "dotenv";
import { existsSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

function walkForEnv(startDir: string): string[] {
  const files: string[] = [];
  let dir = startDir;
  for (let i = 0; i < 6; i++) {
    const candidate = resolve(dir, ".env");
    if (existsSync(candidate)) files.push(candidate);
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return files;
}

/** Load the nearest `.env` files (example folder, then `examples/.env`). */
export function loadExampleEnv(): void {
  const here = dirname(fileURLToPath(import.meta.url));
  for (const file of [...walkForEnv(process.cwd()), ...walkForEnv(here)].reverse()) {
    config({ path: file, override: false });
  }
}

export function requireOpenAIKey(): string {
  const key = process.env.OPENAI_API_KEY;
  if (!key) {
    throw new Error(
      "Set OPENAI_API_KEY. Copy examples/.env.example to examples/.env.",
    );
  }
  return key;
}

File: examples/python/shared/src/lemma_examples_shared/env.py
from __future__ import annotations

import os
from pathlib import Path


def load_example_env() -> None:
    """Load the nearest `.env` files (example folder, then examples/.env)."""
    seen: set[Path] = set()
    for start in (Path.cwd(), Path(__file__).resolve()):
        directory = start if start.is_dir() else start.parent
        for _ in range(6):
            candidate = directory / ".env"
            if candidate not in seen and candidate.is_file():
                seen.add(candidate)
                _apply_env_file(candidate)
            if directory.parent == directory:
                break
            directory = directory.parent


def _apply_env_file(path: Path) -> None:
    for raw_line in path.read_text().splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        key = key.strip()
        value = value.strip().strip("'").strip('"')
        if key and key not in os.environ:
            os.environ[key] = value


def require_openai_key() -> str:
    key = os.environ.get("OPENAI_API_KEY")
    if not key:
        raise RuntimeError(
            "Set OPENAI_API_KEY. Copy examples/.env.example to examples/.env."
        )
    return key
